project-2-Set-Game/SetGame.swift

```diff
@@ -0,0 +1,278 @@
+                    if selectedCards.count == 3 {
+                        selectedCardsMatched = checkIfSelectedcCardsMatch(cards: selectedCards)
+                        if !isGameOver {
+                            score += Int(calculateFactor(isThereAMatch: selectedCardsMatched!))
+                        }
+                        if selectedCardsMatched! {
+                            stopTimers()
+                            opponentLostRound()
+                        }
+                    }
+                }
+            }
+        }
+        checkIfThereIsAMatchOnBoard()
+    }
+    private func calculateFactor(isThereAMatch: Bool) -> Double {
+        let timePassedSinceGameStarted = Double (Date.init().timeIntervalSince(timeWhenGameStarted))
+        return isThereAMatch ?  2000  / timePassedSinceGameStarted : -100
+    }
+    private func replaceMatchedCards(chosenCards: Array<Card>){
+        for card in chosenCards {
+            if let indexOfCard = cardsOnBoard.index(of: card) {
+                // if there are more cards in the deck
+                if deck.count > 0 {
+                    let newCard = deck.remove(at: 0)
+                    cardsOnBoard[indexOfCard] = newCard
+                } else {
+                    cardsOnBoard.remove(at: indexOfCard)
+                }
+            }
+        }
+        selectedCardsMatched = false
+    }
+    private func checkIfSelectedcCardsMatch(cards: Array<Card>) -> Bool {
+        var attributesMapsArray = [[Int: Bool](),[Int: Bool](),[Int: Bool](),[Int: Bool]()]
+        for card in cards {
+            var attributeType = 0
+            for attribute in card.attributes {
+                if attributesMapsArray[attributeType][attribute] == nil {
+                    attributesMapsArray[attributeType][attribute] = true
+                }
+                attributeType += 1
+            }
+        }
+        for attributeMap in attributesMapsArray {
+            if attributeMap.count  == 2 {
+                return false
+            }
+        }
+        return true
+    }
+    public func checkIfThereIsAMatchOnBoard() -> (arrayOfMatchedCards: Array<Int>?,isMatched: Bool){
+        var possibleMatch: Array<Int>? = nil
+        for firstIndex in cardsOnBoard.indices {
+            for secondIndex in firstIndex + 1..<cardsOnBoard.count {
+                for thirdIndex in secondIndex + 1..<cardsOnBoard.count {
+                    if checkIfSelectedcCardsMatch(
+                        cards:[cardsOnBoard[firstIndex],
+                               cardsOnBoard[secondIndex],
+                               cardsOnBoard[thirdIndex]]) {
+                        possibleMatch = [firstIndex,secondIndex,thirdIndex]
+                        return (possibleMatch,true)
+                    }
+                }
+            }
+        }
+        if deck.count == 0 {
+            gameOver()
+        }
+        return (nil,false)
+    }
+    private func dealTwelveCards() {
```

Comment:
dealTwelveCards -> dealMoreCards
dealMoreCards will get a number of cards as argument.
don't forget to check that the deck is not empty or board is full of cards